torch/onnx/verify.py

```diff
@@ -0,0 +1,262 @@
+import copy
+import io
+import os
+import tempfile
+from contextlib import ExitStack
+from typing import Optional, Union
+
+import numpy as np
+
+import torch
+from torch import Tensor
+from torch.onnx.utils import unpack_quantized_tensor
+
+_ORT_PROVIDERS = ["CPUExecutionProvider"]
+
+
+def _flatten_tuples(elem):
+    flattened = []
+    for t in elem:
+        if isinstance(t, tuple):
+            flattened.extend(_flatten_tuples(t))
+        else:
+            flattened.append(t)
+    return flattened
+
+
+def _to_numpy(elem):
+    if isinstance(elem, Tensor):
+        if elem.requires_grad:
+            return elem.detach().cpu().numpy()
+        else:
+            return elem.cpu().numpy()
+    elif isinstance(elem, (list, tuple)):
+        return [_to_numpy(inp) for inp in elem]
+    elif isinstance(elem, bool):
+        return np.array(elem, dtype=bool)
+    elif isinstance(elem, int):
+        return np.array(elem, dtype=int)
+    elif isinstance(elem, float):
+        return np.array(elem, dtype=float)
+    elif isinstance(elem, dict):
+        flattened = []
+        for k in elem:
+            flattened += [_to_numpy(k)] + [_to_numpy(elem[k])]
+        return flattened
+    return elem
+
+
+def _convert_to_onnx(
+    model,
+    model_f: Union[str, io.BytesIO],
+    input=None,
+    opset_version=None,
+    do_constant_folding=True,
+    keep_initializers_as_inputs=True,
+    dynamic_axes=None,
+    input_names=None,
+    output_names=None,
+    fixed_batch_size=False,
+    training=None,
+    verbose=False,
+    ort_providers=_ORT_PROVIDERS,
```

Comment:
_ORT_PROVIDERS is a list which should not be a default value. Should we change _ORT_PROVIDERS to a tuple? https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Reply:
hmm I feel this is acceptable, given that `_ORT_PROVIDERS` is named as constant. And ort_providers is not expected to be mutated. I don't think changing to tuple, or changing to None and setting `ort_providers` inside the function would help very much against unlikely adversarial usage while increasing complexity.

Reply:
I would worry less about adversarial usage than general good practices and what the code tells the readers. If we pass a list as a default, typical readers will wonder if there is anything special about this argument, and whether `_ORT_PROVIDERS` is expected to be mutated. As far as I can tell it is immutable. It is good to just make it a tuple to (1) clarify the intention and (2) avoid the anti-pattern of supplying a mutable as default